### ISSUE TITLE
Suggestion: Generate a page with "Success" for all requests sent to captive.apple.com to trigger the iOS captive portal pop-up

### ISSUE THREAD

**Issue by laozi999** [wifiphisher/wifiphisher#524]
I might be able to do this later tonight.

I'm not sure if it requires any special headers, but I know a public network with a captive portal to test this on.

**Comment by laozi999** [comment]
Would be nice to include a custom command in case you don't want this to occur on iOS & OSX, Android or Windows.

**Comment by Splinter0** [comment]
@laozi999 I love your idea, please keep working on it

**Comment by laozi999** [comment]
what kind of command should we use?

Something like this?

`--showpopup "android,microsoft,apple"`
or
`-sP "android,microsoft,apple"`

**Comment by JetseVerschuren** [comment]
I like the 
`-sP "android,microsoft,apple"`

**Comment by Splinter0** [comment]
@JetseVerschuren Yeah i like that one as well, if you need help just ask @laozi999 :+1:

**Comment by JetseVerschuren** [comment]
Mabe you do
`-sP "all"`

**Comment by laozi999** [comment]
I made some progress on this

I will try to replicate the original response headers as much as possible:

![screenshot from 2017-03-20 11-07-50](https://cloud.githubusercontent.com/assets/24898383/24119377/04aaf0be-0e05-11e7-9702-dc26d9041cf6.png)

![screenshot from 2017-03-20 11-13-01](https://cloud.githubusercontent.com/assets/24898383/24119382/06f92232-0e05-11e7-9960-afe35299654d.png)

Will do more research into captive portal checks of different OS.

**Comment by laozi999** [comment]
I've added custom response headers for apple devices. I will add the other ones soon.

I will replicate the original response headers as much as possible:

Apple:
![screenshot from 2017-03-21 06-14-42](https://cloud.githubusercontent.com/assets/24898383/24120163/b7c5e760-0e07-11e7-9d4a-88e46b608842.png)

Android:
```
HTTP/1.1 307 Temporary Redirect
X-Frame-Options: deny
Cache-Control: no-cache
Location: http://...
Content-Type: text/html
Pragma: no-cache
Expires: 0
Continue: close
Transfer-Encoding: chunked
Server: lighttpd/1.4.39
```

**Comment by Splinter0** [comment]
@laozi999 Good job, can't wait to see more progress :+1:

**Comment by blackHatMonkey** [comment]
@laozi999 Maybe instead of using `--showpopup` we can add `--nopopup`. By this I mean showing pop ups the default and not showing them an option.

**Comment by laozi999** [comment]
@blackHatMonkey based off the scenario that I made (device-modal), there might be an instance where the user doesn't want this popup to show on certain devices.

I think having the freedom to choose is always the better option :)

**Comment by laozi999** [comment]
One second thought I think `--nopopup <devictype>` is better since it is standard practice for captive portals to trigger this popup.

**Comment by sophron** [comment]
@laozi999 What exactly is this feature about?

**Comment by laozi999** [comment]
@sophron It's about the pop-up that is triggered when connecting to an AP with a captive portal. This would enable the user to enable/disable this feature (by changing the response headers sent to `captive.apple.com`, `gstatic.com/generate`_204, etc.)

**Comment by sophron** [comment]
How exactly is this feature working? Whenever a device connects to a WLAN with a captive portal, does it send a request to `captive.apple.com`? Does it expect a response?

**Comment by laozi999** [comment]
Yes, apple devices expect a 202 "OK" response from apple with "Success" as its body to know that there is internet connection. Android expects a 204 "No Content" response with some other set headers. Otherwise responses like the ones I posted above will trigger the pop-up.

**Comment by sophron** [comment]
I see. 

We don't need to provide an option for that. This sounds like something that we want in all cases, correct?

**Comment by JetseVerschuren** [comment]
To trigger a captive portal the device first needs to succeed in a dns request. Then it requests a http page and if you redirect that it triggers a captive portal. So the thing we need is or a upstream connection or a built in dns server.

**Comment by anakin1028** [comment]
@m0rpheus1337

The following is my comprehension for the mechanism of triggering CaptivePortal in Wifiphisher:

Wifiphisher will use `iptable` to redirect the `DNS port 53/tcp/udp` to dnsmasq and redirect the http requests to the tornado server.

For the Android devices it will expect that the response for the certain http request (i.e. `http://clients3.google.com/generate_204`) should have status code `204`, but `tornado web server` will response with `200`; thus triggering CaptivePortal.
